plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TemporalTimeTravelUtil.java

```diff
@@ -0,0 +1,170 @@
+
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogParser.readLastCheckpoint;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogDir;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogJsonEntryPath;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.getEntriesFromJson;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.loadNewTail;
+import static io.trino.spi.StandardErrorCode.INVALID_ARGUMENTS;
+import static java.lang.String.format;
+
+public final class TemporalTimeTravelUtil
+{
+    private TemporalTimeTravelUtil() {}
+
+    public static long findLatestVersionUsingTemporal(TrinoFileSystem fileSystem, String tableLocation, long epochMillis)
+            throws IOException
+    {
+        Optional<LastCheckpoint> lastCheckpoint = readLastCheckpoint(fileSystem, tableLocation);
+        long entryNumber = lastCheckpoint.map(LastCheckpoint::version).orElse(0L);
+
+        String transactionLogDir = getTransactionLogDir(tableLocation);
+
+        Optional<TransactionLogEntries> checkpointEntries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        if (checkpointEntries.isEmpty()) {
+            throw new MissingTransactionLogException(getTransactionLogJsonEntryPath(transactionLogDir, entryNumber).toString());
+        }
+
+        Optional<CommitInfoEntry> commitInfo = checkpointEntries.get().getEntries(fileSystem)
+                .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                .filter(Objects::nonNull)
+                .findFirst();
+
+        // We can not derive useful info from the last check point
+        if (commitInfo.isEmpty()) {
+            return findLatestVersionFromWholeTransactions(fileSystem, tableLocation, epochMillis);
+        }
+
+        if (commitInfo.get().timestamp() == epochMillis) {
+            return entryNumber;
+        }
+
+        if (commitInfo.get().timestamp() < epochMillis) {
+            long tail = searchTowardsTail(fileSystem, entryNumber + 1, transactionLogDir, epochMillis);
+            if (tail >= 0) {
+                return tail;
+            }
+            return entryNumber;
+        }
+        else {
+            long head = searchTowardsHead(fileSystem, entryNumber - 1, transactionLogDir, epochMillis);
+            if (head >= 0) {
+                return head;
+            }
+        }
+
+        throw new TrinoException(INVALID_ARGUMENTS, format("No temporal version history at or before %s", Instant.ofEpochMilli(epochMillis)));
+    }
+
+    private static long searchTowardsTail(TrinoFileSystem fileSystem, long entryNumber, String transactionLogDir, long epochMillis)
+            throws IOException
+    {
+        long version = -1;
+        Optional<TransactionLogEntries> entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
```

Comment:
isn't binary search better solution here? entries are monotonically ordered

Reply:
Here is my understanding: 
there are condition that the commitInfo in the transaction not exists, in this case we can not determine which direction we should check.
For instance, consider the case there are 4 version, [0, 1(empty), 2, 3], the target could one in 0, 2, 3, we first locate the version 1.